Use kni-installer to deploy masters

stbenjam opened the pull request:
Combined with https://github.com/openshift-metalkube/kni-installer/pull/25, will deploy the masters using kni-installer instead of the shell script.

The hardware information is provided to the installer via a flattened version of ironic_nodes.json. I'm not super happy about how that looks, but the limitation of terraform (see https://github.com/openshift-metalkube/kni-installer/pull/25/files#diff-ff320d88a7ec5e5d76645cd4b4a1c28e) in handling structured data doesn't give us a ton of options.

derekhiggins commented:
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/341/

derekhiggins commented:
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/346/

hardys commented:
Just testing this, note for anyone else trying that the kni-installer PR must be rebased to pick up the latest mdns changes or the masters won't come up.

I guess we'll have to land this and the kni-installer patch at the same time to ensure CI continues working - @derekhiggins do we have any way to pull in dependent changes to ensure all works OK when we do that?

stbenjam commented on 03_ocp_repo_sync.sh:
```diff
@@ -30,7 +30,7 @@ function sync_go_repo_and_patch {
 }
 
-sync_go_repo_and_patch github.com/openshift-metalkube/kni-installer https://github.com/openshift-metalkube/kni-installer.git
+sync_go_repo_and_patch github.com/openshift-metalkube/kni-installer https://github.com/openshift-metalkube/kni-installer.git https://github.com/openshift-metalkube/kni-installer/pull/25.patch
```
This applies the kni-installer PR temporarily, to get a good CI run hopefully

stbenjam commented:
Rebased both PR's.  There's a hack in this PR in 03 to patch the kni-installer (see comment above), so we can at least see if CI runs successfully.  

I did have an issue on Friday -- I'm not sure kni-installer will successfully complete.  I removed the early exit (https://github.com/openshift-metalkube/kni-installer/pull/25/files#diff-9ffecdd580b2cb3a1e0e5c2fc92faccfL107), and some operators don't come up like ingress the installer actually fails (very late in the progress - the masters themselves are deployed fine).

derekhiggins commented:
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/369/

hardys commented:
> I did have an issue on Friday -- I'm not sure kni-installer will successfully complete. I removed the early exit (https://github.com/openshift-metalkube/kni-installer/pull/25/files#diff-9ffecdd580b2cb3a1e0e5c2fc92faccfL107), and because some operators don't come up like ingress the installer actually fails (very late in the progress - the masters themselves are deployed fine).

Yeah I did wonder about this - I've seen the same locally so I wonder if we should just move the early exit instead of removing it?

We know the ingress stuff is TODO atm, I think @yboaron is working on it, so we'll need a way to make progress without that for now.

yboaron commented:
@hardys, Yep, I'm working on adding support for the ingress.
I need that the ingress controller will be deployed successfully. 

My plan is to have the router pods on masters-nodes (host networking),
add support for Ingress VIP (VRRP between masters), and wildcard external DNS
entry point to the ingress VIP.

derekhiggins commented:
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/379/

hardys commented:
Sorry needs a rebase but otherwise lgtm, tested locally and seems to work fine, so I'd say lets rebase then merge the kni-installer patch, and this one right after?

derekhiggins commented:
> I guess we'll have to land this and the kni-installer patch at the same time to ensure CI continues working - @derekhiggins do we have any way to pull in dependent changes to ensure all works OK when we do that?

Afraid not, sorry.

stbenjam commented:
Rebased - I removed the hack for the patch as well, so this will fail CI.  I do want to test locally on my box real quick end-to-end, make sure I didnt botch the rebase

derekhiggins commented:
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/411/

stbenjam commented:
It installed:

![image](https://user-images.githubusercontent.com/429763/55161922-fe6db100-513c-11e9-9b2d-fb02a0313b06.png)

derekhiggins commented:
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/412/

derekhiggins commented:
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/413/

stbenjam commented:
Better - should be good to go 

![image](https://user-images.githubusercontent.com/429763/55166278-3678f200-5145-11e9-8d8d-bfee7b46ec86.png)

stbenjam commented:
Rebased, and it does deploy the masters.  I'm stuck at the moment with the ingress controller not coming up but it sounds like https://github.com/openshift-metalkube/dev-scripts/pull/222 fixes that and not related to this PR

![image](https://user-images.githubusercontent.com/429763/55413767-014f1400-5538-11e9-8782-255740c0a766.png)

derekhiggins commented:
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/487/

stbenjam commented:
^ That was the CI from this morning, so encouraging news.  Doesn't cover the latest rebase (although nothing significantly changed)

hardys commented:
Ok tested locally and merged #222, so I think if CI passes here we're good to land the kni-installer patch, remove it from this PR then merge this

derekhiggins commented:
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/489/

stbenjam commented:
Ok, that failure was without #222, so I've rebased it.

derekhiggins commented:
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/492/

derekhiggins commented:
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/498/

hardys commented:
I retriggered CI, hopefully https://github.com/openshift-metalkube/kni-installer/pull/33 landing will help

markmc commented:
This all looks pretty good and I'm keen to see it merged ASAP, leaving any improvements to further iterations. I'm going to play with this for a bit myself now.

The main thing that needs more thought and work as a future iteration is the way we passed the nodes data in install-config.yaml - @stbenjam calls this out in the dev doc too

derekhiggins commented:
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/503/

stbenjam commented:
@hardys Passed! Just removed the patch, these should be good to merge
